Q4/Q8 Tiled Gemm Optimization.

shalinib-ibm opened the pull request:
This patch implemenrts tiled GEMM for large blocks where we pack blocks of 64x64 and perfrom matmul.

10 ~ 30 % improvement in llama-bench and llama-batched-bench with Meta-Llama3-8B Qunatized models( Q4_0 and Q8_0).

*Make sure to read the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

shalinib-ibm committed:
Q4/Q8 Tiled Gemm Optimization.

This patch implemenrts tiled GEMM for large blocks
where we pack blocks of 64x64 and perfrom matmul.

30 ~ 50 % improvement in llama-bench and llama-batched-bench
with Meta-Llama3-8B Qunatized models( Q4_0 and Q8_0).

Signed-off-by: Shalini Salomi Bodapati <Shalini.Salomi.Bodapati@ibm.com>

shalinib-ibm commented:
@ggerganov Can you please review this PR?

ggerganov commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -119,6 +119,30 @@ inline float32x4_t mul(float32x4_t x, float32x4_t y) { return vec_mul(x, y); }
+    }
+}
+static bool key_created = false;
+
```
It would be better to avoid dynamic allocations - none of the code currently uses those. The mechanism for this is to use the `wdata` from `ggml_compute_params` to store scratch data. You'll need to reserve the worst-case `wsize` for your case.

shalinib-ibm replied:
Hi @ggerganov 
I’ve added a note explaining that the params->wdata approach didn’t provide benefits.
When you have some time, could you please take another look at the patch?
Thank you!

ggerganov replied:
@shalinib-ibm For my understanding, is the performance now as good as with the previous approach?

shalinib-ibm committed:
Remove dymanic memory allocation during rutime

This commit addresses review comments.
Also, we have saperated out legacy mnpack path
and matmul_tiled paths for tinyBLAS_Q0_PPC class.

10 ~ 30% improvement in PP Speed with Q4_0 and Q8_0 Models.
Tested with Meta-Llama3-8B quatized models with llama-bench,
llama-batched-bench.

Signed-off-by: Shalini Salomi Bodapati <Shalini.Salomi.Bodapati@ibm.com>

shalinib-ibm commented:
@ggerganov I have addressed your review comment. This patch does not use dynamic allocation now. 
Kindly review the patch. Thanks in advance for your time.

taronaeo commented:
> @taronaeo Can you please review this patch?

I can review this tomorrow, and can only review it at a high-level. I have no PPC hardware to test this on.

shalinib-ibm commented:
> > @taronaeo Can you please review this patch?
> 
> I can review this tomorrow, and can only review it at a high-level. I have no PPC hardware to test this on.
@taronaeo  These results of llama-bench with for Meta-Llama3-8b Q4 model and granite4.0-h-mico Q8 Model on ppc64le linux power10 box.
Similar gains are observed with llama-batched-bench as well.  

/build_base/bin/llama-bench -m /home/shalini/Models/granite-4.0-h-micro-Q8_0.gguf -p 16,32,64,128,256 -n 1 
| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |            pp16 |         70.53 ± 0.12 |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |            pp32 |         74.75 ± 0.83 |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |            pp64 |         75.59 ± 0.28 |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |           pp128 |         75.43 ± 0.32 |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |           pp256 |         75.24 ± 0.18 |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |             tg1 |         18.80 ± 0.04 |

build: 0bcb40b48 (6833)
 ./build_patch_8cc/bin/llama-bench -m /home/shalini/Models/granite-4.0-h-micro-Q8_0.gguf -p 16,32,64,128,256 -n 1 
| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |            pp16 |         77.21 ± 0.22 |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |            pp32 |         86.02 ± 0.09 |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |            pp64 |         90.40 ± 0.04 |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |           pp128 |         96.08 ± 0.06 |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |           pp256 |         98.53 ± 0.01 |
| granitehybrid 3B Q8_0          |   3.16 GiB |     3.19 B | CPU        |      10 |             tg1 |         18.81 ± 0.03 |

build: 962d4e985 (6836)
 ./build_base/bin/llama-bench -m /home/shalini/Models/Meta-Llama-3-8B/ggml-model-q4.gguf -p 16,32,64,128,256 -n 1 
| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |            pp16 |         37.83 ± 0.02 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |            pp32 |         38.90 ± 0.01 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |            pp64 |         39.11 ± 0.01 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |           pp128 |         38.83 ± 0.01 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |           pp256 |         37.84 ± 0.00 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |             tg1 |         13.61 ± 0.02 |

build: 0bcb40b48 (6833)
 ./build_patch_8cc/bin/llama-bench -m /home/shalini/Models/Meta-Llama-3-8B/ggml-model-q4.gguf -p 16,32,64,128,256 -n 1 
| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |            pp16 |         40.81 ± 0.03 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |            pp32 |         44.73 ± 0.01 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |            pp64 |         47.09 ± 0.01 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |           pp128 |         48.20 ± 0.04 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |           pp256 |         47.10 ± 0.00 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | CPU        |      10 |             tg1 |         13.84 ± 0.02 |

build: 962d4e985 (6836)

./llama-cli  -p 'please write a python progarm to print nth fibonacci number' -n 128 -no-cnv -> 
gives 27 t/s for PP on patch while base gives 23 t/s for PP.

ggerganov reviewed (approved):
Make sure you ran perplexity calculations to verify the results are good. Wait for @taronaeo's review

taronaeo reviewed (approved):
LGTM. Just a small spacing fix and we're good to push after CI goes green.

shalinib-ibm committed:
Update ggml/src/ggml-cpu/llamafile/sgemm-ppc.h

small spacing fix

Co-authored-by: Aaron Teo <taronaeo@gmail.com>
